### PR TITLE
detectron2/mmdetection object detection

### PR DESCRIPTION
**Pull request by footprinthere**

### CODE REVIEW

File: examples/detectron2/train.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
+from detectron2.data.datasets import register_coco_instances
```
Comment (ktaebum) [inline examples/detectron2/train.py:1]: Also, please write the original file's path

File: examples/detectron2/train.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        super().__init__(cfg)
+        
+        # Register hook to use Periflow
+        checkpointer = PeriflowCheckpointer(
+            model=self._trainer.model,
+            save_dir=self.cfg.OUTPUT_DIR,
+            trainer=weakref.proxy(self),
+        )
+        hook = PeriflowHook(checkpointer, save_interval)
+        self.register_hooks([hook])
```
Comment (ktaebum) [inline examples/detectron2/train.py:59]: Do we have to define `PeriflowTrainer`? 
I think
```
        # Register hook to use Periflow
        checkpointer = PeriflowCheckpointer(
            model=self._trainer.model,
            save_dir=self.cfg.OUTPUT_DIR,
            trainer=weakref.proxy(self),
        )
        hook = PeriflowHook(checkpointer, save_interval)
        self.register_hooks([hook])
```
could be placed in `main` function

File: examples/detectron2/train.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+
+class PeriflowCheckpointer(DetectionCheckpointer):
```
Comment (ktaebum) [inline examples/detectron2/train.py:62]: naming: `Periflow -> PeriFlow`

File: examples/detectron2/train.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+    def save(self, name, **kwargs) -> None:
+        super().save(name, **kwargs)
+        pf.upload_checkpoint()
+
+
+class PeriflowHook(HookBase):
```
Comment (ktaebum) [inline examples/detectron2/train.py:69]: naming: `Periflow -> PeriFlow`

File: examples/detectron2/train.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        )
+        hook = PeriflowHook(checkpointer, save_interval)
+        self.register_hooks([hook])
+
+
+class PeriflowCheckpointer(DetectionCheckpointer):
+
+    def save(self, name, **kwargs) -> None:
+        super().save(name, **kwargs)
+        pf.upload_checkpoint()
```
Comment (ktaebum) [inline examples/detectron2/train.py:66]: Do we have to define a new `Checkpointer`?
I think detection internally creates a checkpointer? https://github.com/facebookresearch/detectron2/blob/main/detectron2/engine/defaults.py#L450

**Review by ktaebum — COMMENTED** [review]
Thanks! the followings are comments for detectron2 example

File: examples/detectron2/train.py (hunk 1 above)
Comment (footprinthere) [reply examples/detectron2/train.py:7]: Actually there is no 'original file' in the official `detectron2` repo. instead I added a link to a web page that I refered to while writing codes.

**Comment by footprinthere** [comment]
All comments reflected.

File: examples/mmdetection/train.py [hunk 6]
```diff
@@ -0,0 +1,420 @@
...
+        optimizer_config = OptimizerHook(**cfg.optimizer_config)
+    else:
+        optimizer_config = cfg.optimizer_config
+
+    # register hooks
+    pf_training_hook = PeriflowTrainingHook()
+    pf_checkpoint_hook = PeriflowCheckpointHook(
+        interval=1,
+        out_dir=args.checkpoint_dir,
+        create_symlink=False,
```
Comment (ktaebum) [inline examples/mmdetection/train.py:225]: Now, we do not have to enforce `create_symlink=False`

File: examples/mmdetection/train.py [hunk 7]
```diff
@@ -0,0 +1,420 @@
...
+    def before_train_iter(self, runner):
+        super().before_train_iter(runner)
+        pf.start_step()
+
+    def after_train_iter(self, runner):
+        super().after_train_iter(runner)
+        pf.end_step()
+
+
+class PeriflowCheckpointHook(CheckpointHook):
```
Comment (ktaebum) [inline examples/mmdetection/train.py:140]: Our custom checkpoint hook should deal with both `by_epoch` and `not by_epoch` (save by step interval)

File: examples/mmdetection/train.py [hunk 8]
```diff
@@ -0,0 +1,420 @@
...
+            '--options and --cfg-options cannot be both '
+            'specified, --options is deprecated in favor of --cfg-options')
+    if args.options:
+        warnings.warn('--options is deprecated in favor of --cfg-options')
+        args.cfg_options = args.options
+
+    return args
+
+
+class PeriflowTrainingHook(Hook):
```
Comment (ktaebum) [inline examples/mmdetection/train.py:130]: naming: `PeriFlowTrainingHook`

File: examples/mmdetection/train.py (hunk 7 above)
Comment (ktaebum) [inline examples/mmdetection/train.py:140]: naming: `PeriFlowCheckpointHook`

**Review by ktaebum — COMMENTED** [review]
Comments for mmcv

**Comment by footprinthere** [comment]
All comments reflected.

File: examples/mmdetection/train.py [hunk 9]
```diff
@@ -0,0 +1,425 @@
...
+    def after_train_epoch(self, runner):
+        super().after_train_epoch(runner)
+
+        if self.every_n_epochs(runner, self.interval) or \
+         (self.save_last and self.is_last_epoch(runner)):
+            pf.upload_checkpoint()
+
+    def after_train_iter(self, runner):
+        super().after_train_iter(runner)
+
```
Comment (ktaebum) [inline examples/mmdetection/train.py:149]: need
```
if self.by_epoch:
    return
```??
`self.interval` is used in both `by_epoch` mode and `by_iter` mode.

File: examples/mmdetection/train.py [hunk 10]
```diff
@@ -0,0 +1,425 @@
...
+    def after_train_iter(self, runner):
+        super().after_train_iter(runner)
+        pf.end_step()
+
+
+class PeriFlowCheckpointHook(CheckpointHook):
+    def after_train_epoch(self, runner):
+        super().after_train_epoch(runner)
+
```
Comment (ktaebum) [inline examples/mmdetection/train.py:142]: need
```
if not self.by_epoch:
    return
```??
`self.interval` is used in both `by_epoch` mode and `by_iter` mode.

**Comment by footprinthere** [comment]
Comments reflected. Thanks!

**Comment by ktaebum** [comment]
@footprinthere Thanks!

**Pull request merged**